### ISSUE TITLE
4.19.128 boots to black screen

### ISSUE THREAD

**Issue by anon8675309** [linux-surface/linux-surface#216]
The 4.19.128-surface kernel isn't booting properly, it goes to a black screen and has to be hard powered off.

### Environment
- Hardware model: Surface Pro 6
- Kernel version: 4.19.128
- Distribution: Debian 10 (Buster)

<details><summary>Failure to fully boot in 4.19.128, requesting help debugging this.</summary>

I tried alt+F1-F8 to see if it was a problem with X, no effect.  I tried removing the "quiet" and it scrolls messages very quickly and in just over 2 seconds, it's getting to the handover to init before the screen goes blank (I recorded a video so I could see those message, as they scroll by too quickly to read).

Disabling secure boot doesn't make any difference, nor does adding kernel options of "single" or "single init=/bin/bash" in an attempt to drop to a console.  I really don't understand why that didn't work, given the kernel boot messages I saw previously.

I tried compiling 4.19.128 from source, signing it, and booting that.  Same results as the kernel from the linux-surface repo: blank screen.

Booting 5.7.4 works as expected (both the one from the linux-surface repo as well as one I compiled and signed myself).

</details>

**Comment by kitakar5525** [comment]
I tried kernel `4.19.128-surface-lts` with SB1 on Ubuntu 20.04 (booted from external USB storage) but I can't reproduce this issue. Also I don't have issue on Arch Linux.

Try the following to get dmesg log of previous boot:
1. boot with kernel `4.19.128-surface-lts` and wait about 5min to see if it boots
2. if not booting, power off SP6
3. boot with bootable kernel (e.g. `5.7.4`) then
4. get dmesg log of previous boot:
    `journalctl --dmesg --output=short-monotonic --no-hostname --boot=-1`
    I'm not sure if Debian enabled persistent storage for journald. If the command doesn't work, maybe something like the following works:
    `zcat /var/log/dmesg.1.gz `
    The needed log should contain `Linux version 4.19.128-surface-lts` at the top.

**Comment by anon8675309** [comment]
This was very helpful in debugging, thank you.  The issue seems to be related to the firmware for the graphics card (i915), specifically, that the kernel is failing to load the firmware.  Initially, it was because I didn't have them installed in `/lib/firmware` but installing them did not fix the issue (nor cause any issues with the 5.7.4 kernel).

These appear to be the most relevant lines from /var/log/kern.log (additional context is in the attached files at the end of this comment)
```
[    2.200874] i915 0000:00:02.0: Direct firmware load for i915/kbl_guc_ver9_39.bin failed with error -2
[    2.200875] [drm] GuC: Failed to fetch firmware i915/kbl_guc_ver9_39.bin (error -2)
```

The obvious question is whether these files exist.  They do, and they come from the firmware-misc-nonfree package, as one would expect. 
```bash
$ ls -lh /lib/firmware/i915/kbl_guc_ver9_39.bin
-rw-r--r-- 1 root root 145K Aug 22  2019 /lib/firmware/i915/kbl_guc_ver9_39.bin
$ sha256sum /lib/firmware/i915/kbl_guc_ver9_39.bin
957dcd143c348d9a47e0c36cb49970b278790caff62244dd6e354b200261ec74  /lib/firmware/i915/kbl_guc_ver9_39.bin
$ dpkg -S /lib/firmware/i915/kbl_guc_ver9_39.bin
firmware-misc-nonfree: /lib/firmware/i915/kbl_guc_ver9_39.bin
```

I found an [Arch page](https://wiki.archlinux.org/index.php/Intel_graphics#Enable_GuC_/_HuC_firmware_loading) that mentioned that for newer processors the GPU firmware might be needed and that sometimes it's necessary to add `i915.enable_guc=2` to the kernel parameters.  I tried that, but it resulted in the same black screen, and the same errors in the kernel logs about failing to load the firmware.

It's possible that the firmware loading is not required, but since the crash seems to occur in `drm_mode_config_cleanup` and the kernel modules seem to be drm related, it seems likely.

[kern-4.19.128.log](https://github.com/linux-surface/linux-surface/files/4810276/kern-4.19.128.log)

[firmware-4.19.128.log](https://github.com/linux-surface/linux-surface/files/4810284/firmware-4.19.128.log)

**Comment by anon8675309** [comment]
This was very helpful in debugging, thank you.  The issue seems to be related to the firmware for the graphics card (i915), specifically, that the kernel is failing to load the firmware.  Initially, it was because I didn't have them installed in `/lib/firmware` but installing them did not fix the issue (nor cause any issues with the 5.7.4 kernel).

These appear to be the most relevant lines from /var/log/kern.log (additional context is in the attached files at the end of this comment)
```
[    2.200874] i915 0000:00:02.0: Direct firmware load for i915/kbl_guc_ver9_39.bin failed with error -2
[    2.200875] [drm] GuC: Failed to fetch firmware i915/kbl_guc_ver9_39.bin (error -2)
```

The obvious question is whether these files exist.  They do, and they come from the firmware-misc-nonfree package, as one would expect. 
```bash
$ ls -lh /lib/firmware/i915/kbl_guc_ver9_39.bin
-rw-r--r-- 1 root root 145K Aug 22  2019 /lib/firmware/i915/kbl_guc_ver9_39.bin
$ sha256sum /lib/firmware/i915/kbl_guc_ver9_39.bin
957dcd143c348d9a47e0c36cb49970b278790caff62244dd6e354b200261ec74  /lib/firmware/i915/kbl_guc_ver9_39.bin
$ dpkg -S /lib/firmware/i915/kbl_guc_ver9_39.bin
firmware-misc-nonfree: /lib/firmware/i915/kbl_guc_ver9_39.bin
```

I found an [Arch page](https://wiki.archlinux.org/index.php/Intel_graphics#Enable_GuC_/_HuC_firmware_loading) that mentioned that for newer processors the GPU firmware might be needed and that sometimes it's necessary to add `i915.enable_guc=2` to the kernel parameters.  I tried that, but it resulted in the same black screen, and the same errors in the kernel logs about failing to load the firmware.

It's possible that the firmware loading is not required, but since the crash seems to occur in `drm_mode_config_cleanup` and the kernel firmware line also mentions "drm", it seems likely that this is the culprit.

I looked around in the kernel config (menuconfig -> Drivers -> Graphics -> Intel) and saw some alpha support (DRM_I915_ALPHA_SUPPORT) that I could enable using i915.alpha_support=1.  I tried throwing that on the command line, but it did not make a difference.  It crashed in the same location.

[kern-4.19.128.log](https://github.com/linux-surface/linux-surface/files/4810276/kern-4.19.128.log)

[firmware-4.19.128.log](https://github.com/linux-surface/linux-surface/files/4810284/firmware-4.19.128.log)

**Comment by StollD** [comment]
On the patched 4.19 kernel, the GuC will be enabled by default because it is needed for IPTS. Upstream never enabled it by default, and since IPTS on later kernels cannot use the GuC anymore, we don't edit it's default value anymore (which is probably why 5.7 works fine).

Try booting with `i915.enable_guc=0` to disable the GuC. This could probably fix the black screen, but it will disable the touchscreen.

If that works the next step would be to fix the missing GuC firmware. You checked if the firmware exists on the disk, but you should also check if the firmware was added to the initramfs. The GPU driver is usually one of the first things that get loaded, at that stage the root filesystem is not yet mounted. So the firmware needs to be in the initramfs for the kernel to find it.

My theory would be that you installed the kernel first, which generated an initramfs. Then you installed the firmware package that contains the GuC firmware files. But that package didn't trigger an initramfs rebuild, so the GuC firmware never got added to it. You can rebuild it manually by running `sudo update-initramfs -u -k all`.

**Comment by anon8675309** [comment]
That seems to have been it.  I installed the `4.19.128-surface-lts` kernel, and recompiled my own and they both worked.  Your theory that they were missing from initramfs seems like it was likely spot on.

So in the end, it was user error and the solution was to just make sure the firmware was installed.  Thanks for the help everyone. :smile: 

If anyone else runs into this, hopefully they'll find this ticket, be able to debug, and solve their problem.
